### DISCUSSION TITLE
Treeshaking

### DISCUSSION THREAD

**Discussion by borkdude** [squint-cljs/cherry#36] [General]
Continuation of treeshaking discussion: https://github.com/borkdude/cherry/issues/24#issuecomment-1204233553

**Comment by thheller** [comment]
Here is an example of what kind of optimizations the Closure Compiler will do if you let it.

```js
class Example {
  constructor(foo, bar) {
    this.foo = foo;
    this.bar = bar;
  }

  getFoo() {
    return this.foo;
  }

  getBar() {
    return this.bar;
  }

  doSomething() {
    return this.foo + 1;
  }
}

console.log(new Example(1, 2).doSomething());
```

Running this through `esbuild` with `--minify` (or `terser`) you get

```js
class Example{constructor(o,t){this.foo=o,this.bar=t}getFoo(){return this.foo}getBar(){return this.bar}doSomething(){return this.foo+1}}console.log(new Example().doSomething());
```

Which is exactly what it says on the box. Minified. Nothing more. Maybe there are options to make it do more but I haven't found them.

The Closure Compiler in `ADVANCED` mode via `npx google-closure-compiler --js test.js -O ADVANCED` gives you

```js
console.log(2);
```

I actually had to add the `console.log` since otherwise the Closure Compiler just eliminates everything. I'd say that is substantially different and more than just minified, and also more than Tree Shaking would ever do. Obviously, this is not applicable to library code since nothing of the library would be left. 

I'm not even saying this to defend the Closure Compiler in any way. There are big huge trade-offs you have to make to get all of this. The other authors are correct when making different decisions. Sometimes these optimizations are not worth doing.

I'm just trying to emphasize this since this is something you need to be aware of when talking about `:advanced` output. It'll do way more than just shortening variable names and removing whitespace.

Back to the subject of Tree Shaking and cherry. The Closure Compiler is actually capable of outputting ESM code on its own nowadays. See

- https://github.com/google/closure-compiler/blob/3c7d61a10032d6042288ff18a724005b768965c1/src/com/google/javascript/jscomp/CompilerOptions.java#L2979
- https://github.com/google/closure-compiler/blob/3c7d61a10032d6042288ff18a724005b768965c1/src/com/google/javascript/jscomp/ConvertChunksToESModules.java#L34-L67.

However, so far this isn't capable of creating named exports (AFAICT). It just uses import/export to remove the need for the `$APP` hack I did in `shadow-cljs`. This would probably be the place to add this though. Would be nice if this was actually possible. Might be worth looking into instead of trying to rewrite the output after the fact?

**Reply by borkdude** [reply]
Also pulling @jo-sm into this discussion :)

I'm aware of the sophisticated logic that Closure performs. Cherry isn't a replacement for the whole-program optimizing that Closure does: it's just an extra tool that might enable extra use cases (similar to how babashka, nbb, SCI enabled more use cases on top of the "normal" tooling). The idea is to emit ES6 code that the less sophisticated JS tooling can still tree-shake.

**Comment by thheller** [comment]
FWIW I got curious to see how `esbuild` handles tree-shaking and changing the ESM output shadow-cljs produces might help.

`a.js` with
```
let a = 1;
let b = 2;

export { a, b };
```

`index.js` with
```
import { a } from "./a.js";

console.log(a);
```

does indeed shake the `b` out and yields
```
$ npx esbuild --bundle index.js --minify --tree-shaking=true --format=esm
var o=1;console.log(o);
```

but

```
let a; a = 1;
let b; b = 2;

export { a, b };
```
produces
```
$ npx esbuild --bundle index.js --minify --tree-shaking=true --format=esm
var o;o=1;var e;e=2;console.log(o);
```
So, the `b` is no longer removed.

I guess that confirms that changing the output may indeed help tree-shaking. I'm unsure why they don't support this pattern in the first place, but I guess they are being very conservative and a late-assignment is enough to bail. `webpack` appears to have the same behavior.

I can maybe address this by creating different output, so rewriting after the fact isn't needed.

**Reply by borkdude** [reply]
@thheller Wow, thanks for investigating. I'm very open to trying an experimental branch of shadow.

**Reply by thheller** [reply]
You can try `master`. Just `lein install` it for now.

I'll run some tests myself but the output should now fit what other tools expect.

**Reply by borkdude** [reply]
@thheller Tried it just now. I'm getting this output:

```
$ bb build
------------------------------------------------------------------------------

   WARNING: shadow-cljs not installed in project.
   See https://shadow-cljs.github.io/docs/UsersGuide.html#project-install

------------------------------------------------------------------------------
shadow-cljs - config: /Users/borkdude/dev/cherry/shadow-cljs.edn
shadow-cljs - starting via "clojure"
[:cherry] Compiling ...
Closure compilation failed with 5 errors
--- cljs/core.cljs:11330
Imported symbol "h" in chunk "compiler.js" cannot be assigned
--- cljs/tools/reader/impl/utils.cljs:22
Imported symbol "h" in chunk "compiler.js" cannot be assigned
--- cljs/tools/reader/reader_types.cljs:102
Imported symbol "h" in chunk "compiler.js" cannot be assigned
--- edamame/impl/parser.cljc:150
Imported symbol "h" in chunk "compiler.js" cannot be assigned
--- edamame/impl/parser.cljc:735
Imported symbol "h" in chunk "compiler.js" cannot be assigned
```

in branch `shadow-esm` in cherry.

**Reply by borkdude** [reply]
With `--debug`:

```
$ bb build --debug
------------------------------------------------------------------------------

   WARNING: shadow-cljs not installed in project.
   See https://shadow-cljs.github.io/docs/UsersGuide.html#project-install

------------------------------------------------------------------------------
shadow-cljs - config: /Users/borkdude/dev/cherry/shadow-cljs.edn
shadow-cljs - starting via "clojure"
[:cherry] Compiling ...
Closure compilation failed with 5 errors
--- cljs/core.cljs:11330
Imported symbol "$JSCompiler_prototypeAlias$$" in chunk "compiler.js" cannot be assigned
--- cljs/tools/reader/impl/utils.cljs:22
Imported symbol "$JSCompiler_prototypeAlias$$" in chunk "compiler.js" cannot be assigned
--- cljs/tools/reader/reader_types.cljs:102
Imported symbol "$JSCompiler_prototypeAlias$$" in chunk "compiler.js" cannot be assigned
--- edamame/impl/parser.cljc:150
Imported symbol "$JSCompiler_prototypeAlias$$" in chunk "compiler.js" cannot be assigned
--- edamame/impl/parser.cljc:735
Imported symbol "$JSCompiler_prototypeAlias$$" in chunk "compiler.js" cannot be assigned
```

**Reply by borkdude** [reply]
Similar errors with nbb which also uses target esm.

**Reply by jo-sm** [reply]
Yeah, FWIW `esbuild` is generally very conservative about its output and leaves the more involved minimization and dead code elimination to other tools. Running the output from `esbuild ` through `terser` gives a better result.

`module-a.mjs`

```
let a;
a = 1;
let b;
b = 2;

export { a, b };
```

`module-b.mjs`

```
let a;
a = 3;
let b;
b = 4;

export { a, b };
```

`index.mjs`

```
import { a } from "./module-a.mjs";
import { a as b } from "./module-b.mjs";

console.log(a, b);
```

Initial bundling:

```
> ./node_modules/.bin/esbuild --bundle index.mjs --format=esm
// module-a.mjs
var a;
a = 1;
var b;
b = 2;

// module-b.mjs
var a2;
a2 = 3;
var b2;
b2 = 4;

// index.mjs
console.log(a, a2);
```

This isn't really minimized output (as expected, and this is probably due to `esbuild` being very conservative and preferring other tools to handle this). 

Piping this output to `terser` does do proper dead code elimination, though:

```
> ./node_modules/.bin/esbuild --bundle index.mjs --format=esm | ./node_modules/.bin/terser --module --compress
console.log(1,3);
```

**Reply by thheller** [reply]
> Imported symbol "h" in chunk "compiler.js" cannot be assigned

These are errors in the Closure Compiler. They happen if I remove my export-related code entirely. I'm guessing the ES_MODULES output mode has some issues. I'll see if I can make a smaller repro and report upstream.

Guess I'll go back to the hacky version from before for now.

**Reply by borkdude** [reply]
Let me know if/when you're posting a repro, I'll subscribe to the issue.

**Reply by thheller** [reply]
Try `master`. As far as I can tell it seems to help a little bit with my test code. Not sure how it looks in actual projects.

**Reply by borkdude** [reply]
I see a test about it here:

https://github.com/google/closure-compiler/blob/9391ff378af157f1c3fbc836be32f73824e1cf4b/test/com/google/javascript/jscomp/ConvertChunksToESModulesTest.java#L252

It seems they are testing assignments to imports. But I'm not sure how the test example reflects that. Perhaps their DSL builds a module which exports a var a and another module which imports it and assigns it.

**Reply by borkdude** [reply]
@thheller Trying. It seems to progress further, but my build hook is failing:

> Hook [0 shadow.cljs.build-report/hook] failed in stage :flush
> IllegalArgumentException: Cannot open <nil> as a Reader.

I'll remove the build hook for now.

**Reply by borkdude** [reply]
Compilation seems to have worked. I'll do some more testing.

**Reply by borkdude** [reply]
@thheller It seems the output is valid (apart from the build hook failing). Although tree-shaking isn't doing that well yet:

```
$ cat tree.cljs
(ns tree)

(js/console.log (js-keys #js {:a 1}))
```

```
$ cat tree.mjs
import { js_keys } from 'cherry-cljs/cljs.core.js'
console.log(js_keys(({ "a": 1 })));

export {  }
```

```
$ npx esbuild tree.mjs --bundle --minify --platform=node --outfile=dist/index.mjs --format=esm --tree-shaking=true

  dist/index.mjs  282.0kb
```

```
$ ls -la lib
total 1160
drwxr-xr-x   5 borkdude  wheel     160 Aug  6 19:02 .
drwxr-xr-x  40 borkdude  wheel    1280 Aug  6 19:09 ..
-rw-r--r--   1 borkdude  wheel    9391 Aug  6 19:02 cli.js
-rw-r--r--   1 borkdude  wheel  327946 Aug  6 19:02 cljs_core.js
-rw-r--r--   1 borkdude  wheel  246092 Aug  6 19:02 compiler.js
```

Perhaps a step in the right direction though. Or maybe esbuild isn't trying hard enough...

**Reply by thheller** [reply]
Is there a difference in sizer compared to previous versions? I don't expect it to fully eliminate `cljs.core`, there is just too much going on there.

**Reply by borkdude** [reply]
With the previous shadow:

```
$ npx esbuild tree.mjs --bundle --minify --platform=node --outfile=dist/index.mjs --format=esm --tree-shaking=true


  dist/index.mjs  290.8kb
```

**Reply by borkdude** [reply]
Btw, in nbb I didn't get the report error and all tests seemed to pass with the new changes.

**Reply by thheller** [reply]
Hmm so the changes help a little bit I guess.

I can't reproduce your build report failure. Do you have a stacktrace for that?

**Reply by borkdude** [reply]
Note that I commented out the build hooks:

```
#_#_:build-hooks [(shadow.cljs.build-report/hook
                  {:output-to "report.html"})]
```

I'll try again with them enabled again now.

**Reply by borkdude** [reply]
When enabling that again, I get:

```
 $ bb build
------------------------------------------------------------------------------

   WARNING: shadow-cljs not installed in project.
   See https://shadow-cljs.github.io/docs/UsersGuide.html#project-install

------------------------------------------------------------------------------
shadow-cljs - config: /Users/borkdude/dev/cherry/shadow-cljs.edn
shadow-cljs - starting via "clojure"
[:cherry] Compiling ...
Hook [0 shadow.cljs.build-report/hook] failed in stage :flush
IllegalArgumentException: Cannot open <nil> as a Reader.
	clojure.java.io/fn--11641 (io.clj:288)
	clojure.java.io/fn--11641 (io.clj:288)
	clojure.java.io/fn--11530/G--11519--11537 (io.clj:69)
	clojure.java.io/reader (io.clj:102)
	clojure.java.io/reader (io.clj:86)
	clojure.core/apply (core.clj:669)
	clojure.core/slurp (core.clj:7009)
	clojure.core/slurp (core.clj:7009)
	shadow.cljs.build-report/generate-html/fn--20573 (build_report.clj:209)
	shadow.cljs.build-report/generate-html (build_report.clj:209)
	shadow.cljs.build-report/generate-html (build_report.clj:205)
	shadow.cljs.build-report/hook (build_report.clj:346)
	shadow.cljs.build-report/hook (build_report.clj:314)
	clojure.lang.Var.applyTo (Var.java:705)
	clojure.core/apply (core.clj:669)
	clojure.core/apply (core.clj:662)
	shadow.build/configure-hooks-from-config/fn--15906/fn--15914/fn--15915 (build.clj:259)
	shadow.build/execute-hooks/fn--15850/fn--15851 (build.clj:140)
	shadow.build/execute-hooks/fn--15850 (build.clj:136)
	clojure.lang.PersistentArrayMap.kvreduce (PersistentArrayMap.java:429)
	clojure.core/fn--8525 (core.clj:6908)
	clojure.core/fn--8525 (core.clj:6888)
	clojure.core.protocols/fn--8257/G--8252--8266 (protocols.clj:175)
	clojure.core/reduce-kv (core.clj:6919)
	clojure.core/reduce-kv (core.clj:6910)
	shadow.build/execute-hooks (build.clj:133)
	shadow.build/execute-hooks (build.clj:132)
	shadow.build/process-stage (build.clj:168)
	shadow.build/process-stage (build.clj:152)
	shadow.build/flush (build.clj:543)
	shadow.build/flush (build.clj:537)
	shadow.cljs.devtools.api/release* (api.clj:340)
	shadow.cljs.devtools.api/release* (api.clj:324)
	shadow.cljs.devtools.cli-actual/do-build-command (cli_actual.clj:28)
	shadow.cljs.devtools.cli-actual/do-build-command (cli_actual.clj:25)
	shadow.cljs.devtools.cli-actual/do-build-commands (cli_actual.clj:49)
	shadow.cljs.devtools.cli-actual/do-build-commands (cli_actual.clj:38)
	shadow.cljs.devtools.cli-actual/main/body-fn--18465--auto----19174 (cli_actual.clj:166)
	shadow.cljs.devtools.cli-actual/main (cli_actual.clj:165)
	shadow.cljs.devtools.cli-actual/main (cli_actual.clj:132)
	clojure.core/apply (core.clj:671)
	clojure.core/apply (core.clj:662)
	shadow.cljs.devtools.cli-actual/-main (cli_actual.clj:219)
	shadow.cljs.devtools.cli-actual/-main (cli_actual.clj:217)
	clojure.lang.Var.applyTo (Var.java:705)
	clojure.core/apply (core.clj:667)
	clojure.core/apply (core.clj:662)
	shadow.cljs.devtools.cli/-main (cli.clj:75)
	shadow.cljs.devtools.cli/-main (cli.clj:67)
	clojure.lang.Var.applyTo (Var.java:705)
	clojure.core/apply (core.clj:667)
	clojure.main/main-opt (main.clj:514)
	clojure.main/main-opt (main.clj:510)
	clojure.main/main (main.clj:664)
	clojure.main/main (main.clj:616)
	clojure.lang.Var.applyTo (Var.java:705)
	clojure.main.main (main.java:40)
```

**Reply by thheller** [reply]
Ah ok, yeah that is just because of `lein install` not creating all the other necessary files. I happen to have them so they get bundled in the `.jar`. Won't be a problem in a proper release then.

**Reply by borkdude** [reply]
@thheller Cool. Is there a way to get notifications of new shadow releases? I see you don't use github releases for example.

**Reply by borkdude** [reply]
Btw, just tried closure on the bundled output from esbuild:

```
npx google-closure-compiler --module_resolution NODE --compilation_level ADVANCED  --js=dist/index.mjs  --js_output_file=dist/out.js --language_in ECMASCRIPT_2020
```

It spits out 3k warnings, but the output works and is much smaller:

```
$ ls -lat dist
total 1096
-rw-r--r--   1 borkdude  wheel  167548 Aug  7 13:33 out.js
```

**Reply by thheller** [reply]
Well, that is expected. Should also be true when just running the closure compiler alone, without running `esbuild` first. `ADVANCED` will still do its thing properly with the extra information provided by the `tree.mjs`. You could also try what `terser` does with this?

I just pushed another update to `master` which seems to cover most cases. Turns out using their `ES_MODULES` output just required one additional setting. I'll do a few more tests and then make a release. I do not expect the changes to help `esbuild` much but maybe it does.

**Reply by borkdude** [reply]
Thanks. I can test with master in an hour or two.

**Reply by borkdude** [reply]
@thheller Got some time to test now. I'm getting lots of 

```
Imported symbol "g_" in chunk "nbb_reagent_dom_server.js" cannot be assigned
--- reagent/dom/server.cljs:23
```

etc again now when compiling nbb.

**Reply by thheller** [reply]
Can you test with pseudo-names? Otherwise its very hard to figure out what `h_` might be.

**Reply by thheller** [reply]
Hmm, yeah I can reproduce this. These all appear to be CLJS dynamic vars used in `binding`. I'll see if I can figure something out.

**Comment by borkdude** [comment]
I’ll be back in an hour. Meanwhile you can try the nbb repo with the
shadow-esm branch if you wish.

On Sun, 7 Aug 2022 at 15:45, Thomas Heller ***@***.***> wrote:

> Can you test with pseudo-names? Otherwise its very hard to figure out what
> h_ might be.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/borkdude/cherry/discussions/36#discussioncomment-3343430>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACFSBUBO6LQUYQLMBXUPGLVX64ZHANCNFSM55PWJAXA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
https://www.michielborkent.nl
https://www.eetvoorjeleven.nu
